src/wiki/decorators.py

```diff
@@ -34,8 +36,41 @@ def response_forbidden(request, article, urlpath, read_denied=False):
         )
 
 
+def which_article(path=None, article_id=None, **kwargs):
```

Comment:
Since this is nested in the `decorators` module and isn't a decorator, it would be better to mark it internal, for instance renaming it to `_which_article`.

Another thing might be that the name is a bit odd. Perhaps just `_get_article` is good for now, simply suggesting that it's the internal part of the `get_article` decorator?

Reply:
The reason I extracted this function is because it is useful in other places – I needed it for understanding where links lead in redlinks and in whatlinkswhere. I would move it out of decorators into urls or util, but I wanted to keep it in a single-file diff for easier comparison for this PR.

Reply:
I've just noticed that there is also a `response_forbidden` function mis-placed in the `decorators` module, which is also not really a decorator, so the strictness would be a bit over-kill.

Reply:
As I said, this was done with an eye towards transparency. I'm actually in favour of cleaning it up, but making it explicit that it's a cleanup, not part of this restructuring.